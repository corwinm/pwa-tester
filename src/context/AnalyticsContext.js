import ReactGA from "react-ga";

export function initializeGA() {
  ReactGA.initialize("UA-139919526-1", {
    testMode: process.env.NODE_ENV !== "production",
  });
}

export function pageView(path) {
  ReactGA.pageview(process.env.PUBLIC_URL + path);
}
