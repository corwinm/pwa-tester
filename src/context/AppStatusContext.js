import React from "react";
import useAppStatus from "custom-hooks/useAppStatus";

export const AppStatusContext = React.createContext({
  updateAvailable: false,
  offLineReady: false,
  registration: null,
  installPrompt: null
});

export const AppStatusProvider = ({ children }) => {
  const appUpdateAvailable = useAppStatus();
  return (
    <AppStatusContext.Provider value={appUpdateAvailable}>
      {children}
    </AppStatusContext.Provider>
  );
};
