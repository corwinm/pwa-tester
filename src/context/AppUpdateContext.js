import React from "react";

export const AppUpdateContext = React.createContext({
  updateAvailable: false,
  offLineReady: false,
  registration: null,
  installPrompt: null
});
