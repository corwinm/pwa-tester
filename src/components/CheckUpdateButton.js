import React, { useContext, useState } from "react";
import { AppStatusContext } from "context/AppStatusContext";

export default function CheckUpdateButton() {
  const { registration, updateAvailable } = useContext(AppStatusContext);
  const [updating, setUpdating] = useState(false);

  if (!registration) return null;

  if (updateAvailable) {
    const updateServiceWorker = () => {
      registration.waiting.postMessage({type: "SKIP_WAITING"});
    };
    return (
      <button className="button is-info" onClick={updateServiceWorker}>
        <span className="icon">
          <i className="fas fa-sync" />
        </span>
        <span>Update Now</span>
      </button>
    );
  }

  async function checkForUpdate() {
    setUpdating(true);
    const newRegistration = await registration.update();
    console.log(newRegistration);
    setUpdating(false);
  }

  return (
    <button className="button is-info" onClick={checkForUpdate}>
      <span className="icon">
        <i className={"fas fa-sync" + (updating ? " fa-spin" : "")} />
      </span>
      <span>{updating ? "Updating" : "Check for Update"}</span>
    </button>
  );
}
