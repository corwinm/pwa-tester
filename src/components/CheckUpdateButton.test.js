import React from "react";
import { render, cleanup, fireEvent, wait, act } from "react-testing-library";
import CheckUpdateButton from "./CheckUpdateButton";
import { AppStatusContext } from "context/AppStatusContext";

afterEach(cleanup);

test("matches unregistered snapshot", () => {
  const { container } = render(
    <AppStatusContext.Provider value={{ registration: null }}>
      <CheckUpdateButton />
    </AppStatusContext.Provider>
  );
  expect(container.firstChild).toMatchSnapshot();
});

test("matches registered snapshot", () => {
  const { container } = render(
    <AppStatusContext.Provider value={{ registration: {} }}>
      <CheckUpdateButton />
    </AppStatusContext.Provider>
  );
  expect(container.firstChild).toMatchSnapshot();
});

test("matches update available snapshot", () => {
  const { container } = render(
    <AppStatusContext.Provider
      value={{ registration: {}, updateAvailable: true }}
    >
      <CheckUpdateButton />
    </AppStatusContext.Provider>
  );
  expect(container.firstChild).toMatchSnapshot();
});

test("clicking update now posts message to registration", () => {
  const mockRegistration = { waiting: { postMessage: jest.fn() } };
  const { getByText } = render(
    <AppStatusContext.Provider
      value={{ registration: mockRegistration, updateAvailable: true }}
    >
      <CheckUpdateButton />
    </AppStatusContext.Provider>
  );

  fireEvent.click(getByText(/update now/i));

  expect(mockRegistration.waiting.postMessage.mock.calls.length).toBe(1);
  expect(mockRegistration.waiting.postMessage).toBeCalledWith("skipWaiting");
});

test("clicking check for update updates the registration", async () => {
  const mockRegistration = { update: jest.fn() };

  const { getByText } = render(
    <AppStatusContext.Provider
      value={{ registration: mockRegistration, updateAvailable: false }}
    >
      <CheckUpdateButton />
    </AppStatusContext.Provider>
  );

  mockRegistration.update.mockImplementation(async () => {
    await wait(() => expect(getByText(/updating/i)).toBeTruthy());

    return { newregistration: "new registration" };
  });

  fireEvent.click(getByText(/Check for Update/i));

  expect(mockRegistration.update).toBeCalled();

  return await wait(() => expect(getByText(/Check for Update/i)).toBeTruthy());
});
