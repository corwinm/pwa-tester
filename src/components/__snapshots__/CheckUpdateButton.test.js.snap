// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches registered snapshot 1`] = `
<button
  class="button is-info"
>
  <span
    class="icon"
  >
    <i
      class="fas fa-sync"
    />
  </span>
  <span>
    Check for Update
  </span>
</button>
`;

exports[`matches unregistered snapshot 1`] = `null`;

exports[`matches update available snapshot 1`] = `
<button
  class="button is-info"
>
  <span
    class="icon"
  >
    <i
      class="fas fa-sync"
    />
  </span>
  <span>
    Update Now
  </span>
</button>
`;
